catkin_ws/src/state_estimation/src/state_aggregator.py

```diff
@@ -43,57 +67,95 @@ def __init__(self):
+        q_auv = self.q_dvl_mount.inverse()*q_dvl
+
+        # quaternion of AUV in world frame
+        self.q_auv_dvl = self.q_world.inverse()*q_auv
+
+        # position of DVL (and also AUV TODO) in NED frame
+        pos_dvl = np.array([data.x, data.y, data.z]) 
+
+        # AUV position/offset from pos_world (relative to global frame)
+        pos_auv = pos_dvl - self.pos_world
 
-    def dr_cb(self,data):
-        self.x = data.x - self.x_offset
-        self.y = data.y - self.y_offset
+        # AUV position in world frame
+        self.pos_auv_dvl = quaternion.rotate_vectors(self.q_world, pos_auv)
+
+        # update overall state
+        # TODO - check about views of arrays in numpy
+        self.pos_auv[0:2] = self.pos_auv_dvl[0:2]
 
-    def dvl_offset(self, data):
-        self.x_offset = self.x 
-        self.y_offset = self.y
 
     def imu_cb(self, imu_msg):
-        # quaternion in global (imu) frame
+        # quaternion of imu in NED frame
         q_imu = imu_msg.quaternion
         q_imu = np.quaternion(q_imu.w, q_imu.x, q_imu.y, q_imu.z) 
 
-        # quaternion in world frame
-        self.q_auv= self.q_world.inverse()*q_imu
+        # quaternion of AUV in NED frame
+        # sensor is mounted, may be oriented differently from AUV axes
+        q_auv = self.q_imu_mount.inverse()*q_imu
+
+        # quaternion of AUV in world frame
+        self.q_auv_imu = self.q_world.inverse()*q_auv
+
+        # update overall state
+        self.q_auv = self.q_auv_imu
         self.update_euler()
 
 
     def depth_sensor_cb(self, depth_msg):
-        # TODO - have depth microcontroller publish ready value
-        self.z = depth_msg.data
+        # TODO - check does the 0 depth coincide with NED frame?
+        # position of depth sensor (and assumed AUV TODO) in NED? frame
+        pos_depth = np.array([0.0, 0.0, depth_msg.data]) 
+
+        # position/offset from pos_world (relative to global orientation)
+        pos_auv = pos_depth - self.pos_world
+
+        # position in world frame
+        self.pos_auv_depth_sensor = quaternion.rotate_vectors(self.q_world, pos_auv)
+        
+        # update overall state - TODO - when AUV is rotated depth might correspond to other axes 
+        self.pos_auv[2] = self.pos_auv_depth_sensor[2]
+
+
+    def reset_state_cb(self, _):
+        # TODO - for now, only reset the orientation
+        # maybe on reset find best world frame that keeps the same x-y plane?
+        # new world position is current AUV position in global frame
+        # self.pos_world = self.pos_auv + self.pos_world
+        # self.pos_auv = np.array([0.0, 0.0, 0.0])
 
-    def imu_reset_cb(self, _):
         # copy of q_auv in (old) world frame
         q_auv = self.q_auv
         q_auv = np.quaternion(q_auv.w, q_auv.x, q_auv.y, q_auv.z)
 
         # new world quaternion is q_auv in global (imu) frame
-        self.q_world = self.q_world.inverse()*q_auv
+        self.q_world = self.q_world*q_auv
```

Comment:
does this need to be self.imu_mount*self.q_world*q_auv? I'm not sure

Reply:
I don't believe so, we want the world frame to coincide with the auv frame, not the imu frame. The compensation for the imu mounting orientation happens in imu_cb (106-111):
```
# quaternion of AUV in NED frame
# sensor is mounted, may be oriented differently from AUV axes
q_auv = self.q_imu_mount.inverse()*q_imu

 # quaternion of AUV in world frame
self.q_auv_imu = self.q_world.inverse()*q_auv
```

We're always working in world/auv frame q_imu_mount is just a way to get to q_auv by processing the data coming from the imu